StacksAndQueues/animalShelter/solution1.js

```diff
@@ -0,0 +1,184 @@
+            array.push(temp.name);
+            temp = temp.prev;
+        }
+        console.log("CATS", array);
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    dogQueue(name) {
+        let dog = new Pet(name, this.DTail);
+        if (this.DHead == null) {
+            this.DTail = dog;
+            this.DHead = dog;
+        } else {
+            this.DTail.prev = dog;
+            dog.next = this.DTail;
+            this.DTail = dog;
+        }
+        return dog;
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    catQueue(name) {
+        let cat = new Pet(name, this.CTail);
+        if (this.CHead == null) {
+            this.CTail = cat;
+            this.CHead = cat;
+        } else {
+            this.CTail.prev = cat;
+            cat.next = this.CTail;
+            this.CTail = cat;
+        }
+        return cat;
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    enqueue(name, type) {
+        let pet;
+        if (type == "dog") {
+            pet = this.dogQueue(name);
+        } else {
+            pet = this.catQueue(name);
+        }
+        if (this.ACHead == null) {
+            this.ACTail = pet;
+            this.ACHead = pet;
+        } else {
+            pet.acNext = this.ACTail;
+            this.ACTail.acPRev = pet;
+            this.ACTail = pet;
+        }
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    dequeueDog() {
+        let adopted = this.DHead;
+        if (adopted.acNext != null) {
+            adopted.acNext.acPRev = adopted.acPRev;
+        }
+        if (this.ACHead === adopted) {
+            this.ACHead = adopted.acPRev;
+        }
+        if (this.DHead.prev != null) {
+            this.DHead = this.DHead.prev;
+            this.DHead.next = null;
+        } else {
+            this.DHead = null;
+            this.DTail = null;
+        }
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    dequeueCat() {
+        let adopted = this.CHead;
+        if (adopted.acNext != null) {
+            adopted.acNext.acPRev = adopted.acPRev;
+        }
+        if (this.ACHead === adopted) {
+            this.ACHead = adopted.acPRev;
+        }
+        if (this.CHead.prev != null) {
+            this.CHead = this.CHead.prev;
+            this.CHead.next = null;
+        } else {
+            this.CHead = null;
+            this.CTail = null;
+        }
+    }
+
+    // Time complexity is Big O(1);
+    // Space complexity is Big O(1);
+    dequeueAny() {
+        if (this.ACHead === this.CHead) {
+            this.dequeueCat();
+        } else {
+            this.dequeueDog();
+        }
+    }
+}
+
+
+const ac = new AdoptionCenter();
+ac.enqueue("D1", "dog");
+ac.enqueue("C1", "cat");
+ac.enqueue("C2", "cat");
+ac.enqueue("D2", "dog");
+console.log(ac.printAnimals()); //ANIMALS [ 'D1', 'C1', 'C2', 'D2' ]
```

Comment:
I think the undefined prints are coming from here because `printAnimals()` does it's own printing and doesn't return anything.

Reply:
Ahhh. That makes sense. So @annamikmorozova it makes sense to me that the console.logs on the bottom that log printAnimals should be removed since printAnimals does it's own console. Thoughts?

Reply:
@truthgoddess sounds good! Fixed!